gdrive_sync/api.py

```diff
@@ -201,7 +202,10 @@ def process_file_result(
 def stream_to_s3(drive_file: DriveFile):
     """ Stream a Google Drive file to S3 """
     try:
-        s3 = boto3.resource("s3")
+        if settings.ENVIRONMENT == "dev":
```

Comment:
This was test related.  I tried using simply `get_boto3_resource` here and the tests would fail.  Here's an example of one of the errors:

```
    @pytest.mark.django_db
    @pytest.mark.parametrize("num_errors", [2, 3, 4])
    def test_stream_to_s3_error(settings, mocker, num_errors):
        """Task should mark DriveFile status as failed if an s3 upload error occurs more often than retries"""
        settings.ENVIRONMENT = "test"
        settings.CONTENT_SYNC_RETRIES = 3
        should_raise = num_errors >= 3
        mocker.patch("gdrive_sync.api.GDriveStreamReader")
        mock_boto3 = mocker.patch("gdrive_sync.api.boto3")
        mock_boto3.resource.return_value.Bucket.side_effect = [
            *[HTTPError() for _ in range(num_errors)],
            mocker.Mock(),
        ]
        drive_file = DriveFileFactory.create()
        if should_raise:
            with pytest.raises(HTTPError):
                api.stream_to_s3(drive_file)
        else:
>           api.stream_to_s3(drive_file)

gdrive_sync/api_test.py:189: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
content_sync/decorators.py:29: in wrapper
    return func(*args, **kwargs)
gdrive_sync/api.py:220: in stream_to_s3
    bucket.upload_fileobj(
/usr/local/lib/python3.8/site-packages/boto3/s3/inject.py:580: in bucket_upload_fileobj
    return self.meta.client.upload_fileobj(
/usr/local/lib/python3.8/site-packages/boto3/s3/inject.py:540: in upload_fileobj
    return future.result()
/usr/local/lib/python3.8/site-packages/s3transfer/futures.py:106: in result
    return self._coordinator.result()
/usr/local/lib/python3.8/site-packages/s3transfer/futures.py:265: in result
    raise self._exception
/usr/local/lib/python3.8/site-packages/s3transfer/tasks.py:255: in _main
    self._submit(transfer_future=transfer_future, **kwargs)
/usr/local/lib/python3.8/site-packages/s3transfer/upload.py:554: in _submit
    if not upload_input_manager.requires_multipart_upload(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <s3transfer.upload.UploadSeekableInputManager object at 0x7f2bdd06a310>
transfer_future = <s3transfer.futures.TransferFuture object at 0x7f2bdd06fcd0>, config = <boto3.s3.transfer.TransferConfig object at 0x7f2bdd50a550>

    def requires_multipart_upload(self, transfer_future, config):
>       return transfer_future.meta.size >= config.multipart_threshold
E       TypeError: '>=' not supported between instances of 'MagicMock' and 'int'

/usr/local/lib/python3.8/site-packages/s3transfer/upload.py:240: TypeError
```

I think this is another example of Mocker not behaving.  In in `test_stream_to_s3` we are using `mock_boto3 = mocker.patch("gdrive_sync.api.boto3")` and I don't think that our call to initialize boto3 is properly being mocked because we aren't calling boto3 directly in the scope of the test function.  It seems if the function is called indirectly, `mocker.patch` doesn't work.  I even tried changing it to `mock_boto3 = mocker.patch("main.s3_utils.boto3")` which would hypothetically patch the `boto3` instance being pulled in by the utility function being called by the function the test is calling, and I still got that `TypeError: '>=' not supported between instances of 'MagicMock' and 'int'`.  Any ideas?

Reply:
Oh, nevermind on this one.  I didn't restart web container after changing the mock target :facepalm: